src/Microsoft.ML.Data/Transforms/SelectColumnsTransform.cs

```diff
@@ -490,21 +511,59 @@ public int GetInputIndex(int outputIndex)
             {
                 var outputToInputMapping = new List<int>();
                 var columnCount = inputSchema.ColumnCount;
-                int outputIdx = 0;
 
-                for (int colIdx = 0; colIdx < columnCount; ++colIdx)
+                if (keepColumns)
                 {
-                    if (!keepHidden && inputSchema.IsHidden(colIdx))
-                        continue;
+                    // With KeepColumns, the order that is specified is preserved in the mapping.
+                    // For example if a given input has the columns of ABC and the select columns are
+                    // specified as CA, then the output will be CA.
+
+                    // In order to account for keeping hidden columns, build a dictionary of
+                    // column name-> list of column indices. This dictionary is used for
+                    // building the final mapping.
+                    var columnDict = new Dictionary<string, List<int>>();
+                    for(int colIdx = 0; colIdx < inputSchema.ColumnCount; ++colIdx)
+                    {
+                        if (!keepHidden && inputSchema.IsHidden(colIdx))
+                            continue;
+
+                        var columnName = inputSchema[colIdx].Name;
+                        if (columnDict.TryGetValue(columnName, out List<int> columnList))
+                            columnList.Add(colIdx);
+                        else
+                        {
+                            columnList = new List<int>();
+                            columnList.Add(colIdx);
+                            columnDict.Add(columnName, columnList);
+                        }
+                    }
 
-                    var columnName = inputSchema[colIdx].Name;
-                    var selected = selectedColumns.Contains(columnName);
-                    selected = (keepColumns) ? selected : !selected;
-                    if (selected)
+                    // Since the ordering matters, iterate through the selected columns
+                    // finding the associated index that should be used.
+                    foreach(var columnName in selectedColumns)
+                    {
+                        if (columnDict.TryGetValue(columnName, out List<int> columnList))
```

Comment:
>TryGetValue [](start = 39, length = 11)

I think this should fail if the schema doesn't contain a column by that name.

Reply:
Are you referring to in general or specifically in this case? In general there is an ignoreMissing parameter that when it is true will ignore any columns specified not being in the input. This is checked earlier before the Mapper is created.

---
In reply to: [228247404](https://github.com/dotnet/machinelearning/pull/1371#discussion_r228247404) [](ancestors = 228247404)

Reply:
@Zruty0 - Yael makes a great point! What are your thoughts on defaulting ignoreMissing to false?

Reply:
I agree

---
In reply to: [228266038](https://github.com/dotnet/machinelearning/pull/1371#discussion_r228266038) [](ancestors = 228266038)

Reply:
Ok I will default ignoreMissing to false.

Reply:
ignoreMissing now defaults to false.

---
In reply to: [228264779](https://github.com/dotnet/machinelearning/pull/1371#discussion_r228264779) [](ancestors = 228264779,228264118,228257636,228247404)